x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/util/BatchedDataIterator.java

```diff
@@ -0,0 +1,189 @@
+ * @param <E> The collection that should be used (i.e. Set, Deque, etc.)
+ */
+public abstract class BatchedDataIterator<T, E extends Collection<T>> {
+    private static final Logger LOGGER = LogManager.getLogger(BatchedDataIterator.class);
+
+    private static final String CONTEXT_ALIVE_DURATION = "5m";
+    private static final int BATCH_SIZE = 10000;
+
+    private final Client client;
+    private final String index;
+    private volatile long count;
+    private volatile long totalHits;
+    private volatile String scrollId;
+    private volatile boolean isScrollInitialised;
+
+    protected BatchedDataIterator(Client client, String index) {
+        this.client = Objects.requireNonNull(client);
+        this.index = Objects.requireNonNull(index);
+        this.totalHits = 0;
+        this.count = 0;
+        this.isScrollInitialised = false;
+    }
+
+    /**
+     * Returns {@code true} if the iteration has more elements.
+     * (In other words, returns {@code true} if {@link #next} would
+     * return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    public boolean hasNext() {
+        return !isScrollInitialised || count != totalHits;
+    }
+
+    /**
+     * The first time next() is called, the search will be performed and the first
+     * batch will be given to the listener. Any subsequent call will return the following batches.
+     * <p>
+     * Note that in some implementations it is possible that when there are no
+     * results at all, the first time this method is called an empty Collection is given to the listener.
+     */
+    public void next(ActionListener<E> listener) {
+        if (!hasNext()) {
+            listener.onFailure(new NoSuchElementException());
+        }
+        ActionListener<SearchResponse> wrappedListener = ActionListener.wrap(
+            searchResponse -> {
+                scrollId = searchResponse.getScrollId();
+                mapHits(searchResponse, listener);
+            },
+            listener::onFailure
+        );
+        if (scrollId == null) {
+            initScroll(wrappedListener);
+        } else {
+            SearchScrollRequest searchScrollRequest = new SearchScrollRequest(scrollId).scroll(CONTEXT_ALIVE_DURATION);
+            ClientHelper.executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+                ClientHelper.DATA_FRAME_ORIGIN,
+                searchScrollRequest,
+                wrappedListener,
+                client::searchScroll);
+        }
+    }
+
+    private void initScroll(ActionListener<SearchResponse> listener) {
+        LOGGER.trace("ES API CALL: search index {}", index);
+
+        isScrollInitialised = true;
+
+        IndicesOptions indicesOptions = SearchRequest.DEFAULT_INDICES_OPTIONS;
+        indicesOptions = IndicesOptions.fromOptions(true, indicesOptions.allowNoIndices(), indicesOptions.expandWildcardsOpen(),
```

Comment:
Is all this to add `ignore_unavailable`?

Reply:
Yes, getting all the correct options is a pain. I was reading through all the available options here, and it seems that `lenientExpandOpen` does the exact same thing except for the ignoring throttling + throw on closed. Admittedly, in the scenario I am using this, if either of those are the case for the indices in question, we have bigger issues at hand. 

I will simplify to `lenientExpandOpen` for now. If there is push-back, I will try to craft a more complicated option set :D .